src/subtype.c

```diff
@@ -1444,56 +1421,97 @@ static int is_definite_length_tuple_type(jl_value_t *x)
     return k == JL_VARARG_NONE || k == JL_VARARG_INT;
 }
 
-static int forall_exists_equal(jl_value_t *x, jl_value_t *y, jl_stenv_t *e)
-{
-    if (obviously_egal(x, y)) return 1;
-
-    if ((is_indefinite_length_tuple_type(x) && is_definite_length_tuple_type(y)) ||
-        (is_definite_length_tuple_type(x) && is_indefinite_length_tuple_type(y)))
-        return 0;
+static int _forall_exists_subtype(jl_value_t *x, jl_value_t *y, jl_stenv_t *e, int param, int *count, int *noRmore);
 
-    if (jl_is_uniontype(x) && jl_is_uniontype(y)) {
-        // For 2 unions, try a more efficient greedy algorithm that compares the unions
-        // componentwise. If it returns `false`, we forget it and proceed with the usual
-        // algorithm. If it returns `true` we try returning `true`, but need to come back
-        // here to try the usual algorithm if subtyping later fails.
-        jl_unionstate_t *state = &e->Runions;
-        jl_saved_unionstate_t oldRunions; push_unionstate(&oldRunions, state);
-        if (pick_union_decision(e, 1) == 0) {
-            if (equal_unions((jl_uniontype_t*)x, (jl_uniontype_t*)y, e))
-                return 1;
-            pop_unionstate(state, &oldRunions);
+static int obviously_nounion(jl_value_t *x, jl_stenv_t *e, jl_typeenv_t *log) JL_NOTSAFEPOINT
+{
+    if (x == NULL || x == (jl_value_t*)jl_any_type || x == jl_bottom_type)
+        return 1;
+    if (jl_is_unionall(x))
+        return obviously_nounion(((jl_unionall_t *)x)->body, e, log);
+    if (jl_is_datatype(x)) {
+        jl_datatype_t *xd = (jl_datatype_t *)x;
+        for (int i = 0; i < jl_nparams(xd); i++) {
+            jl_value_t *param = jl_tparam(xd, i);
+            if (jl_is_vararg(param))
+                param = jl_unwrap_vararg(param);
+            if (!obviously_nounion(param, e, log))
+                return 0;
         }
+        return 1;
+    }
+    if (!jl_is_typevar(x))
+        return 0;
+    jl_typeenv_t *t = log;
+    while (t != NULL) {
+        if (x == (jl_value_t *)t->var)
+            return 0;
+        t = t->prev;
     }
+    jl_typeenv_t newlog = { (jl_tvar_t*)x, NULL, log };
+    jl_varbinding_t *xb = lookup(e, (jl_tvar_t *)x);
+    return obviously_nounion(xb ? xb->lb : ((jl_tvar_t *)x)->lb, e, &newlog) &&
+           obviously_nounion(xb ? xb->ub : ((jl_tvar_t *)x)->ub, e, &newlog);
```

Comment:
ub and lb also need to be type-equal (such as egal), or there is possibly many unions between those bounds. For example Union<:T<:Any. Or is this just an attempted estimate of the decision points?

I think it might be best to move this commit to a different PR, since I am fairly confident in the rest of this change, but not as clear on this commit change currently.

Reply:
> Or is this just an attempted estimate of the decision points?

This is an attempted estimation.

With out this commit we will consume at least 2 `Runion` for each invariant parameter.
And some packages (e.g. `Latexify.jl`) form complex `Tuple` with thousands of `Pairs`. (see the test added.)

Reply:
Ah, thanks for the explanation. This sounds good to me now. Maybe we could name it `may_contain_union_decision`?